src/main/java/org/mitre/synthea/engine/State.java

```diff
@@ -218,6 +222,11 @@ public boolean process(Person person, long time) {
         person.history = moduleHistory;
         // add this state to history to indicate we returned to this module
         person.history.add(0, this);
+        // start using the current encounter, it may have changed
```

Comment:
I think we have to. In the example unit test (and in the to-be anemia submodule where users discovered this) the current encounter is ended, because the patient needs to be admitted as an inpatient for whatever happens next. When the module returns, whatever other clinical activity remained in the parent module should still be conducted during the inpatient encounter. So, I think it makes sense.

Reply:
Ok. I need some time to think about what other implications this may have. 
I know that the intent of the anemia submodule is that anemia can be secondary to various other conditions so they want to have other modules call it. But if the submodule could change the encounter then you have to be aware of that and be careful to structure the module around that.

Reply:
Well, in principle or spirit this isn't any different than the way all the modules or submodules work today, in that they all have the ability to change anything at any time that could alter the way your current module works. Smoking is intertwined, so are opioid prescriptions. 

Submodules are essentially reusable blocks of code. Some code is idempotent, some is not. Either way, you always have to be careful of how you call submodules or change attributes.

I'm also happy to completely remove modules-by-encounter and only use encounters by time, but that has many implications and would require a different, much more sophisticated pull request.